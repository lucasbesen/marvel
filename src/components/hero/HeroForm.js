import React, { Component } from 'react';
import styled from 'styled-components';
import { Form, Field } from 'react-final-form';
import { toast } from 'react-toastify';
import Button from '@material-ui/core/Button/Button';
import List from '@material-ui/core/List/List';
import ListItem from '@material-ui/core/ListItem/ListItem';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import IconButton from '@material-ui/core/IconButton';
import BackIcon from '@material-ui/icons/ArrowBack';
import ListSubheader from '@material-ui/core/ListSubheader';

import { TextField } from '../../components/common';

const Wrapper = styled.div`
  display: flex;
  flex: 1;
  width: 100%;
  background-color: #fff;
  justify-content: space-between;
  margin-bottom: 20px;
`;

const FieldsWrapper = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  margin-left: 10%;
`;

const FieldWrapper = styled.div`
  display: flex;
  flex: 1;
`;

const Picture = styled.img`
  width: 200px;
  border-radius: 100px;
`;

class HeroForm extends Component {
  onSubmit = values => {
    const { hero } = this.props;

    localStorage.setItem(`${hero.id}-name`, values.name);
    localStorage.setItem(`${hero.id}-description`, values.description);
    toast.success('Saved!');
  };

  validate = values => {
    const errors = {};
    if (!values.name) {
      errors.name = 'Required';
    }
    if (!values.description) {
      errors.description = 'Required';
    }
    return errors;
  };

  render() {
    const { hero, history } = this.props;
    const initialValues = {
      name: localStorage.getItem(`${hero.id}-name`) || hero.name,
      description: localStorage.getItem(`${hero.id}-description`) || hero.description,
    };

    return (
      <Form
        onSubmit={this.onSubmit}
        initialValues={initialValues}
        validate={this.validate}
        render={({ handleSubmit, submitting, pristine }) => (
          <form onSubmit={handleSubmit}>
            <IconButton onClick={() => history.push('/heroes')}>
              <BackIcon />
            </IconButton>
            <Wrapper>
              <Picture src={`${hero.thumbnail.path}.${hero.thumbnail.extension}`} />
              <FieldsWrapper>
                <FieldWrapper>
                  <Field name="name" component={TextField} type="text" label="Name" fullWidth />
                </FieldWrapper>
                <FieldWrapper>
                  <Field name="description" component={TextField} type="text" label="Description" fullWidth />
                </FieldWrapper>
                <div>
                  <Button
                    variant="contained"
                    color="secondary"
                    size="small"
                    type="submit"
                    disabled={submitting || pristine}
                  >
                    Save
                  </Button>
                </div>
              </FieldsWrapper>
            </Wrapper>
            <List subheader={<ListSubheader component="div">Movies/TV Shows</ListSubheader>}>
              {hero.series &&
                hero.series.items.map((item, index) => (
                  <ListItem key={`${item.name}-${index}`}>
                    <ListItemText primary={item.name} />
                  </ListItem>
                ))}
            </List>
          </form>
        )}
      />
    );
  }
}

export default HeroForm;
