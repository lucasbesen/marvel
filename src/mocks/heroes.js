export const heroMock = {
  id: 1,
  name: 'Captain America',
  thumbnail: {
    path: 'www.google.com',
  },
  series: {
    items: [
      {
        name: 'TV Show 1',
      },
      {
        name: 'TV Show 2',
      },
    ],
  },
};
