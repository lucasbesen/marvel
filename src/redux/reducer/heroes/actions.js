import { getHero as getHeroService, listHeroes as listHeroesService } from '../../../services';

export const HERO: string = 'hero:HERO';
export const HEROES: string = 'heroes:HEROES';
export const PAGINATION: string = 'heroes:PAGINATION';

export const fetchHeroes = (limit: string, offset: string, name: string): void => async dispatch => {
  const response = await listHeroesService(limit, offset, name);
  dispatch({ type: HEROES, payload: response.results });
  dispatch({
    type: PAGINATION,
    payload: {
      count: response.count,
      limit: response.limit,
      offset: response.offset,
      total: response.total,
    },
  });
};

export const getHero = (id: string): void => async dispatch => {
  dispatch({ type: HEROES });
  const response = await getHeroService(id);
  dispatch({ type: HERO, payload: response.results[0] });
};
