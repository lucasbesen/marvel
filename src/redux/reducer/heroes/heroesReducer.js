import createReducer from '../createReducer';
import { HERO, HEROES, PAGINATION } from './actions';

import type { State } from '../../../types/State';
import type { Action } from '../../../types/Action';

const initialState: State = {
  hero: {},
  heroes: [],
  pagination: {},
};

const heroesReducer = createReducer(initialState, {
  [HERO]: (state: State, action: Action) => ({
    ...state,
    hero: action.payload,
  }),
  [HEROES]: (state: State, action: Action) => ({
    ...state,
    heroes: action.payload,
  }),
  [PAGINATION]: (state: State, action: Action) => ({
    ...state,
    pagination: action.payload,
  }),
});

export default heroesReducer;
