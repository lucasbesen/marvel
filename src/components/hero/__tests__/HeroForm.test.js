import React from 'react';
import renderer from 'react-test-renderer';

import { heroMock } from '../../../mocks/heroes';
import HeroForm from '../HeroForm';

it('should render HeroForm without error', () => {
  const component = renderer.create(<HeroForm hero={heroMock} history={jest.fn()} />);
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
