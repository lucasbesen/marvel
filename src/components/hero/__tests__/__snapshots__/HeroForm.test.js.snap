// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render HeroForm without error 1`] = `
<form
  onSubmit={[Function]}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
    </span>
  </button>
  <div
    className="sc-bwzfXH rBETs"
  >
    <img
      className="sc-ifAKCX gEGXBp"
      src="www.google.com.undefined"
    />
    <div
      className="sc-htpNat jcigDw"
    >
      <div
        className="sc-bxivhb hpFLOR"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
          >
            Name
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              style={
                Object {
                  "width": 250,
                }
              }
              type="text"
              value="Captain America"
            />
          </div>
        </div>
      </div>
      <div
        className="sc-bxivhb hpFLOR"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Description
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              name="description"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              style={
                Object {
                  "width": 250,
                }
              }
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeSmall Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Save
          </span>
        </button>
      </div>
    </div>
  </div>
  <ul
    className="MuiList-root MuiList-padding MuiList-subheader"
  >
    <div
      className="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      Movies/TV Shows
    </div>
    <li
      className="MuiListItem-root MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          TV Show 1
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          TV Show 2
        </span>
      </div>
    </li>
  </ul>
</form>
`;
